### ISSUE TITLE
Google Image Family doesn't detect when "latest" image changes.

### ISSUE THREAD

**Issue by ldelossa** [hashicorp/terraform#10984]
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

### Terraform Version
Run `terraform -v` to show the version. If you are not running the latest version of Terraform, please upgrade because your issue may have already been fixed.

terraform --version
Terraform v0.7.13
(Not using latest to avoid create_timeout bug)

### Affected Resource(s)
Please list the resources as a list, for example:
- google_compute_instance_template
- google_compute_instance_group_manager

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
```hcl
resource "google_compute_instance_template" "sentry_web" {
  depends_on = ["google_compute_instance.sentry-redis-01",
                "google_compute_instance.sentry-redis-02",
                "google_compute_instance.sentry-postgresql-01",
                "consul_keys.sentry_web"]

  name_prefix = "player-sentry-${var.cluster_name}-web-${var.image_version}-"
  description = "Sentry Web Template for env: ${var.cluster_name} version: ${var.image_version}"
  region = "us-east1"

  machine_type = "n1-standard-2"
  disk {
    source_image = "sentry-web-${var.cluster_name}"
    auto_delete = true
    boot = true
  }

  tags = ["sentry", "sentry-web"]

  network_interface {
    subnetwork = "us-east1-b"
    access_config = {
      // Empty to default to ephemeral public ip
    }
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "google_compute_instance_template" "sentry_web_test" {
  depends_on = ["google_compute_instance.sentry-redis-01",
                "google_compute_instance.sentry-redis-02",
                "google_compute_instance.sentry-postgresql-01",
                "consul_keys.sentry_web"]

  name_prefix = "player-sentry-${var.cluster_name}-web-${var.image_version}-test"
  description = "Sentry Web Template for env: ${var.cluster_name} version: ${var.image_version}"
  region = "us-east1"

  machine_type = "n1-standard-2"
  disk {
    source_image = "sentry-web-${var.cluster_name}"
    auto_delete = true
    boot = true
  }

  tags = ["sentry", "sentry-web"]

  network_interface {
    subnetwork = "us-east1-b"
    access_config = {
      // Empty to default to ephemeral public ip
    }
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "google_compute_instance_template" "sentry_worker_preprocessor" {
  depends_on = ["google_compute_instance.sentry-redis-01",
                "google_compute_instance.sentry-redis-02",
                "google_compute_instance.sentry-postgresql-01",
                "consul_keys.sentry_web"]

  name_prefix = "player-sentry-${var.cluster_name}-preprocessor-${var.image_version}-"
  description = "Sentry Worker Preprocessor Template for env: ${var.cluster_name} version: ${var.image_version}"
  region = "us-east1"

  machine_type = "n1-standard-2"
  disk {
    source_image = "sentry-worker-preprocessor-${var.cluster_name}"
    auto_delete = true
    boot = true
  }

  tags = ["sentry", "sentry-worker-preprocessor"]

  network_interface {
    subnetwork = "us-east1-b"
    access_config = {
      // Empty to default to ephemeral public ip
    }
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "google_compute_instance_template" "sentry_worker_processor" {
  depends_on = ["google_compute_instance.sentry-redis-01",
                "google_compute_instance.sentry-redis-02",
                "google_compute_instance.sentry-postgresql-01",
                "consul_keys.sentry_web"]

  name_prefix = "player-sentry-${var.cluster_name}-processor-${var.image_version}-"
  description = "Sentry Worker Preprocessor Template for env: ${var.cluster_name} version: ${var.image_version}"
  region = "us-east1"

  machine_type = "n1-standard-2"
  disk {
    source_image = "sentry-worker-processor-${var.cluster_name}"
    auto_delete = true
    boot = true
  }

  tags = ["sentry", "sentry-worker-processor"]

  network_interface {
    subnetwork = "us-east1-b"
    access_config = {
      // Empty to default to ephemeral public ip
    }
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "google_compute_instance_template" "sentry_worker_save" {
  depends_on = ["google_compute_instance.sentry-redis-01",
                "google_compute_instance.sentry-redis-02",
                "google_compute_instance.sentry-postgresql-01",
                "consul_keys.sentry_web"]

  name_prefix = "player-sentry-${var.cluster_name}-worker-save-${var.image_version}-"
  description = "Sentry Worker Preprocessor Template for env: ${var.cluster_name} version: ${var.image_version}"
  region = "us-east1"

  machine_type = "n1-standard-2"
  disk {
    source_image = "sentry-worker-save-${var.cluster_name}"
    auto_delete = true
    boot = true
  }

  tags = ["sentry", "sentry-worker-save"]

  network_interface {
    subnetwork = "us-east1-b"
    access_config = {
      // Empty to default to ephemeral public ip
    }
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "google_compute_instance_group_manager" "sentry_web_managed_instance_group" {
  depends_on = ["google_compute_instance_template.sentry_web"]
  name = "player-sentry-${var.cluster_name}-web"
  base_instance_name = "player-sentry-${var.cluster_name}-web-${var.image_version}"
  instance_template = "${google_compute_instance_template.sentry_web.self_link}"
  zone = "us-east1-b"
  target_size = 3
  named_port {
    name = "sentry-web"
    port = 9000
  }
}

resource "google_compute_instance_group_manager" "sentry_worker_preprocessor_managed_instance_group" {
  depends_on = ["google_compute_instance_template.sentry_web"]
  name = "player-sentry-${var.cluster_name}-worker-preprocessor"
  base_instance_name = "player-sentry-${var.cluster_name}-worker-preprocessor-${var.image_version}"
  instance_template = "${google_compute_instance_template.sentry_worker_preprocessor.self_link}"
  zone = "us-east1-b"
  target_size = 8
  named_port {
    name = "sentry-web"
    port = 9000
  }
}

resource "google_compute_instance_group_manager" "sentry_worker_processor_managed_instance_group" {
  depends_on = ["google_compute_instance_template.sentry_web"]
  name = "player-sentry-${var.cluster_name}-worker-processor"
  base_instance_name = "player-sentry-${var.cluster_name}-worker-processor-${var.image_version}"
  instance_template = "${google_compute_instance_template.sentry_worker_processor.self_link}"
  zone = "us-east1-b"
  target_size = 11
  named_port {
    name = "sentry-web"
    port = 9000
  }
}

resource "google_compute_instance_group_manager" "sentry_worker_save_managed_instance_group" {
  depends_on = ["google_compute_instance_template.sentry_web"]
  name = "player-sentry-${var.cluster_name}-worker-save"
  base_instance_name = "player-sentry-${var.cluster_name}-worker-save-${var.image_version}"
  instance_template = "${google_compute_instance_template.sentry_worker_save.self_link}"
  zone = "us-east1-b"
  target_size = 5
  named_port {
    name = "sentry-web"
    port = 9000
  }
}
```

### Debug Output
Please provider a link to a GitHub Gist containing the complete debug output: https://www.terraform.io/docs/internals/debugging.html. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

### Panic Output
If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`.

### Expected Behavior
When google instance template resource is created with disk pointing to a image family, terraform should evaluate latest image in image family. This should decide whether we need to recreate the corresponding instance group manager resource.

### Actual Behavior
When uploading a new image to the image family with packer and then re-running terraform plan, it sees no need to re-create the group manager resource. In my testing, the group manager resource needs to still be re-created (or duplicated) in order for it to point to the latest image in the image family.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Create image with image family in packer
2. Define an instance template which points to the image family
3. Define a instance group manager resource pointing to the instance template
4. Create new packer image which is now latest
5. View in GCE that old image is reference in instance template resource
6. Run terraform apply - should see no need for any actions. 

### Important Factoids
Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? Custom version of OpenStack? Tight ACLs?

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- GH-1234

**Comment by paddyforan** [comment]
Hey @ldelossa, thanks for opening an issue! If I understand correctly, when you upload a new image, Terraform (which is set to use the "latest" image in the family) doesn't see the need to recreate the template.

Is this accurate? Is it accurate that the version you're setting them to is the string "latest" (or similar) and you're not hard-coding specific versions into your terraform config? Sorry, I'm having a hard time telling from the config you pasted above; it seems you have a number of variables in there and didn't include samples of what those would be, so it could really go either way. Any clarification would be helpful and appreciated!

**Comment by ldelossa** [comment]
Ahh sorry about all the variables - after taking a look I realized how confusing that can be lol. The evaluated instance template block would look like this:

```
resource "google_compute_instance_template" "sentry_worker_preprocessor" {
  depends_on = ["google_compute_instance.sentry-redis-01",
                "google_compute_instance.sentry-redis-02",
                "google_compute_instance.sentry-postgresql-01",
                "consul_keys.sentry_web"]

  name_prefix = "player-sentry-prod-preprocessor-1-0-"
  description = "Sentry Worker Preprocessor Template for env: ${var.cluster_name} version: ${var.image_version}"
  region = "us-east1"

  machine_type = "n1-standard-2"
  disk {
    source_image = "sentry-worker-preprocessor-prod"
    auto_delete = true
    boot = true
  }

  tags = ["sentry", "sentry-worker-preprocessor"]

  network_interface {
    subnetwork = "us-east1-b"
    access_config = {
      // Empty to default to ephemeral public ip
    }
  }

  lifecycle {
    create_before_destroy = true
  }
}
```

sentry-worker-preprocessor-prod is the name of the image family. I was not under the impression that I may need 

sentry-worker-preprocessor-prod/latest ?

**Comment by paddyforan** [comment]
No worries! Thanks for the quick clarification.

So, the good news: I've found the source of the bug! When we create a template, if a source image is set, we resolve that to a specific disk image and store _that_ in the template. This is done to try and do the right thing with whatever input the config has, without making people type out long, arcane strings to specify the image.

The bad news: I'm not sure offhand how to go about fixing it without causing some serious backwards incompatibilities or having the implementation be pretty brittle. I'm actually a little surprised our current implementation works--we're specifying a full URL to the disk, which the API docs don't explicitly seem to support.

I'm going to think about this and pick at it a little bit, but the code in question is in this file: https://github.com/hashicorp/terraform/blob/3f8b25dbba2e52a4157cd1b44e2c8409a3cb48e1/builtin/providers/google/image.go

Also worth noting that that function is used for disks and instances, in addition to instance templates, so those resources will show the same behaviour and be impacted by whatever fix we implement.

**Comment by ldelossa** [comment]
@paddyforan hmm I see.

This block of code:

```
// If it doesn't exist, try to see if it works as an image family:
			image, err = c.clientCompute.Images.GetFromFamily(c.Project, name).Do()
			if err == nil {
				return image.SelfLink, nil
			}
```

This should always return you the latest image within the image family correct? If that is the case, then if an 'instance template function' is calling the function you provided above, and a new image was added to the image family just before this function was called, wouldn't it return a NEW link to a NEW image ? And terraform would see that a change in the instance template needs to happen?

**Comment by ldelossa** [comment]
@paddyforan I'm starting to dig into the code now. Any update on this issue? This is kind of a deal breaker for including Terraform into experimental CD pipeline. This breaks terraform refreshing automated images created with Packer in GCE environments.

**Comment by ldelossa** [comment]
Thanks for your response. 

For now that is my current work-around. I'm just going to pass the new images into TF as a variable. Not the worst thing in the world, but obviously using families correctly in TF is easier. It's a great feature that allows you to just forget about image creation for the most part.

I see what you're saying here. I was looking over the code briefly, targeting the 'refresh' step in 'terraform plan'

I was hoping I could find a quick and dirty way to just hook into the provider resource and run a ne resolve, then update state - but I was having difficulty. The options you suggest here do make sense to me. For now I can deal with the work around stated above.

**Comment by paddyforan** [comment]
> As far as my experience with image families and the API, it's completely valid to make all requests with just the image family name.

That matches with my reading of the API docs. The only reason I suggest it isn't recommended behaviour is because the docs also say that instance groups should be homogeneous and should create identical resources, which isn't really the case if the image they're using is mutable. But at the end of the day, the API allows it, and so Terraform probably should, too. I just need to figure out how to get us from where we are to there, safely. :)

**Comment by ldelossa** [comment]
Sounds good, thanks for the speedy responses.

**Comment by paddyforan** [comment]
Hey @ldelossa, just wanted to update you and let you know I haven't forgotten about this. It's bugging me on a personal, vendetta-esque level right now. :) Just wanted to take a moment to spell out the solution I've arrived at, so people can jump in and tell me if I'm wrong. Here's what I'm planning:

* the image resolver now looks for one of the following formats:
  * `family/{family}`
  * `projects/{project}/global/images/{image}`
  * `{project}/{image}`
  * `{project}/{family}`
  * `{family}`
  * `{image}`
  * `{full URL to image, e.g. from compute_image.self_link}`
  * `global/images/{private-image}`
  * `global/images/family/{private-family}`

  This should cover all previously acceptable inputs, as well as add a bunch more. These all (except the full URL, which is passed through untouched) expand to a canonical string, as seen here (under `sourceImage`): https://cloud.google.com/compute/docs/reference/latest/disks

  Because we expand to this canonical string, and not a URL, we are consistent with the API behaviour, of new images being used as they get pushed and a new instance is spun up.
* We add a `compute_image_family` data source that exposes a `latest` property, which uses the old behaviour of pulling in the latest image at the time Terraform was run. Because this is passed as a URL, `resolveImage` passes it transparently to the API, mimicking old behaviour.

This, I think, is the most consistent and least surprising way for the functionality to work. Unfortunately, because of the backwards compatibility concerns, it won't ship until 0.9. Fortunately, the `master` branch is now for 0.9, so as soon as I get it coded up, it should be OK to merge, and people can run against master if necessary until the 0.9 beta/RCs/final release drops.

At the moment, I've got a unit test for the new string parsing behaviour, and just need to figure out the best way to make private images work (need to provision the private image and family before I can test against them), and I've updated `resolveImage` to pass the tests. Next step after that is the new data source. Hopefully I'll be able to update again soon with a PR number. :)

Feel free to weigh in if the proposed solution doesn't fit your needs, and thanks for reporting the issue!

**Comment by paddyforan** [comment]
Oh, I should also mention that because `compute_image_family` would be a data source, it _should_ also create a diff when a new image ships, letting you know that the image has changed and a new apply is needed.

**Comment by ldelossa** [comment]
Okay I believe I'm following you here:

compute_image_family is a data source which will be used to fill in the "image" properties in the HCL that we are writing. This will always return the newest canonical name of the image within the image family. 

The addition of resolving more name spaces seems like a big place. As far as how you're explaining things - it sounds good to me.

**Comment by paddyforan** [comment]
I just opened #12223 to fix the image resolution logic to allow families to be specified, not just images. If it's merged in time, it will be part of 0.9. I'll continue working on the data source for families, to return the latest image whenever `terraform plan` is called. Thanks for reporting the issue, and for all your patience!

**Comment by paddyforan** [comment]
The image resolution logic fix was just released in 0.9. If it's OK, I'm going to close this out and open a new issue to track creating a data source for an image family.

**Comment by ldelossa** [comment]
Works for me. Thanks
